### ISSUE TITLE
[Workflow] Moving workflow configuration into separate file

### ISSUE THREAD

**Issue by Guikingone** [symfony/symfony#21197]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.3

I was using the workflow component and find the configuration approach a little bit to "complicated" by the way that we need to inject key into the config.yml and expect that no "yaml" error are made, this way, i was thinking about injecting a workflows.yml file into the config.yml this way : 

```yaml
imports:
    - { resource: parameters.yml }
    - { resource: security.yml }
    - { resource: services.yml }
    - { resource: workflows.yml }
```

And defining workflow into the workflows.yml file (in the same directory as config.yml) : 

```yaml
# Workflow phase
framework:
    workflows:
        # User process
       user_process:
           type: 'workflow'
           marking_store:
               type: 'multiple_state'
               arguments:
                   - 'currentStatus'
           supports:
               - UserBundle\Entity\User
           places:
              - inscription
              - validation
              - connexion
              - informations
              - full_profile
              - regular_user
           transitions:
               register_phase:
                   from: inscription
                   to: validation
               first_connexion_phase:
                   from: validation
                   to: connexion
               informations_phase:
                   from: connexion
                   to: informations
               full_phase:
                   from: informations
                   to: full_profile
               regular_phase:
                   from: full_profile
                   to: regular_user
```

The data are just for test but this way, we just inject the keys into config.yml and can configure the workflows into a dedicated file. 

Any thoughts about this approach ?

**Comment by lyrixx** [comment]
Hello @Guikingone 

I'm not sure to understand what you want. You can already put the workflow configuration into a dedicated file. (like any other bundle configuration BTW)

See this application for [an example](https://github.com/lyrixx/SFLive-Paris2016-Workflow/blob/master/app/config/config.yml#L5)

**Comment by Guikingone** [comment]
Hi @lyrixx

In fact, i was just thinking about a "native" way of configuring workflow without using the config.yml file, like we can set services and routes inside a dedicated file.

By the actual way, i find (i can be the only one because of the relative youngness of the component) that configuring a workflow can be boring and pretty hard due by the fact that Yaml could be corrupt by spaces and tabulation, i know, moving the logic into a separate file can lead to the same problems but the actual "way" is probably not "dev-friendly" by the way that you need to put the workflow key at the end of the 'framework' key and most of the time, it lead to yaml parse error. 

By using separate file, the dev can configure the workflow process inside a file dedicate to this task and "surcharge" this file inside every bundle if he want like services and route.

I find 'strange' that we need to put the workflow config key into the config file rather than a dedicate file who can be simplier to the dev and the team who work with him, plus, the workflow component is pretty powerfull and can lead to a better structure on many project, with a dedicated file, this can lead to greater use and better understing of what he do.

**Comment by theofidry** [comment]
It makes sense to me at least to specify a config file like for the router for example in `framework.workflow` 👍

**Comment by lyrixx** [comment]
What would be format of the "separate file" ?

**Comment by theofidry** [comment]
```yaml
# app/config.yml
framework:
    workflows: '%kernel.root_dir%/config/workflows.yml'
```

```yaml
# app/config/workflows.yml
       user_process:
           type: 'workflow'
           marking_store:
               type: 'multiple_state'
               arguments:
                   - 'currentStatus'
           supports:
               - UserBundle\Entity\User
           places:
              - inscription
              - validation
              - connexion
              - informations
              - full_profile
              - regular_user
           transitions:
               register_phase:
                   from: inscription
                   to: validation
               first_connexion_phase:
                   from: validation
                   to: connexion
               informations_phase:
                   from: connexion
                   to: informations
               full_phase:
                   from: informations
                   to: full_profile
               regular_phase:
                   from: full_profile
                   to: regular_user
```

That would also allow in the future to allow a XML/PHP file for it if that makes sense

**Comment by lyrixx** [comment]
@theofidry  I don't understand why you want to invent a new way of doing something that is already possible with the same exact configuration format.

Then, you can already import .xml file. or .php file (like for every other part of the framework configuration).

**Comment by Guikingone** [comment]
As i see in the doc, you must use the default structure to load workflows : 

```yaml
framework:
    workflows:
        blog_publishing:
            type: 'workflow' # or 'state_machine'
            marking_store:
                type: 'multiple_state' # or 'single_state'
                arguments:
                    - 'currentPlace'
            supports:
                - AppBundle\Entity\BlogPost
            places:
                - draft
                - review
                - rejected
                - published
            transitions:
                to_review:
                    from: draft
                    to:   review
                publish:
                    from: review
                    to:   published
                reject:
                    from: review
                    to:   rejected
```

This way, the workflows key is right under the framework key but that's not (in my opinion) the best way to load this type of config, is you look at the services or security files, they use a 'unique' key who allow to load file without using the framework key, the solution of @theofidry could be great but the component become 'dependant' of the framework key, for exemple, we can use this type of convention : 

```yaml
workflows:
    user_process:
           type: 'workflow'
           marking_store:
               type: 'multiple_state'
               arguments:
                   - 'currentStatus'
           supports:
               - UserBundle\Entity\User
           places:
              - inscription
              - validation
              - connexion
              - informations
              - full_profile
              - regular_user
           transitions:
               register_phase:
                   from: inscription
                   to: validation
               first_connexion_phase:
                   from: validation
                   to: connexion
               informations_phase:
                   from: connexion
                   to: informations
               full_phase:
                   from: informations
                   to: full_profile
               regular_phase:
                   from: full_profile
                   to: regular_user
```

Like the framework key, the workflow key can be 'primary' key and define into a new file to be load into config.yml without using file loading with the kernel.root.dir params.

**Comment by lyrixx** [comment]
The DIC and the Router component have their own configuration loader system
The security have its own key because it is not in the `FrameworkBundle` (so `framework` key) but in the `SecurityBundle` (so `security` key)

So just to be sure, you just want to get ride of the `framework` key?

**Comment by Guikingone** [comment]
Yes, it's just about forgetting the framework key and/or moving the logic of workflow config into a dedicated file (if the framework key can be forget, the move of the logic into a separate file isn't obligatory but could be a plus), i think that could help to the configuration of the component and could lead to less parse error.

**Comment by wouterj** [comment]
@Guikingone these "keys" are related to specific Dependency Injection Extensions (those of the FrameworkBundle and SecurityBundle).

It is perfectly possible to put this config in a seperate file. Read the [How to Organize Configuration Files](http://symfony.com/doc/current/configuration/configuration_organization.html) documentation article for more information about this.

 > By the actual way, i find (i can be the only one because of the relative youngness of the component) that configuring a workflow can be boring and pretty hard due by the fact that Yaml could be corrupt by spaces and tabulation, i know, moving the logic into a separate file can lead to the same problems but the actual "way" is probably not "dev-friendly" by the way that you need to put the workflow key at the end of the 'framework' key and most of the time, it lead to yaml parse error.

This is actually more a problem of the Yaml syntax: It's "human-friendliness" is soo overdone that it actually becomes pretty complex. However, you are free to use the XML or PHP format to configure any DI extension in Symfony.

**Comment by theofidry** [comment]
@lyrixx personally it's not so much the framework key that bother me, it's that for the framework config, everything is in one place. As the workflow config grows, it becomes very noisy and it's better to separate it, but again if you do that right now, you no longer have all of your framework config in one place which is not ideal.

Granted it's not too big of an issue really, but I believe it's a small addition that could be appreciated. That said, I'm not a heavy user of the workflow component, so maybe someone like @Nyholm could provide a more relevant opinion on this.

**Comment by lyrixx** [comment]
Again and again, it's **already** possible to put the configuration into another file. So if the issue is not about the framework key, it don't understand this issue.

**Comment by theofidry** [comment]
@lyrixx let me rephrase: the `framework` key becomes a "problem" only when you move the workflow config in a separate file.

**Comment by wouterj** [comment]
I think what this issue is about:

Instead of including the workflow config in the DI extension, the component should provide it's own YamlFileLoader, XmlFileLoader and PhpFileLoader.

The benefit of this is, is that it might makes configuration a bit nicer when using the component standalone. However, as this introduces quite some extra work, I propose to not do that:

 * Using the `DefinitionBuilder`, configuring the component already is quite nice in standalone usage
 * I'm not very sure how often this component is/will be used outside Symfony

**Comment by Guikingone** [comment]
👍 for @theofidry vision, ths key become problematic when you move the logic and i find that the logic of the workflow could/should be move into a separate file, this configuration could be pretty large and hard to set into the same file as the configuration for the whole framework. 

@wouterj Yes, the component isn't very used outside of the framework but that's not really the problem, problem is, the config file could be pretty large and big and with a new 'secondary' key inside this file, this could lead to a harder reading level.

**Comment by lyrixx** [comment]
Could you fork the SE and show me what you want?

**Comment by wouterj** [comment]
Well, it's not much of a harder reading level imo. After all, what we're talking about is 1 word ("framework") and 4 spaces of indentation (which you can perfectly reduce to 1 space btw) per line.

What this still makes very clear is that the workflow is part of the FrameworkBundle and one has to look at the Framework DI extension in order to see how it's processed.

**Comment by Guikingone** [comment]
@lyrixx Yes, i do that tomorrow when i come home.

@wouterj Yes, the process is pretty long and the investisment probably don't justify the work but i find strange this type of config, i know, we can inject workflows file into the config file (i take a look at your repo @lyrixx) but i just find that with the configuration of the framework and the workflow, this could lead to misunderstand of all the key and harder time to read the file.

**Comment by Nyholm** [comment]
I agree with that the configuration under the `framework` could be massive. Especially with the workflow component. What I've done (and what have been suggested in this PR) is that you put it in a separate file.

```yaml
# config.yml
framework: 
  secret: foobar
  default_locale:  en
  # ...
```
```yaml
# workflows.yml
framework:
  workflows: 
    my_workflow: 
      # ...
```

I do not think there is a problem with starting the "workflows.yml" with the `framework:` key. In fact, it is only logical. 🖖

**Comment by lyrixx** [comment]
The initial request was about declaring the configuration in another file and then referencing this file into the `config.yml` file. As I said it is already supported.

Then, it looks like there is an issue with the first `framework` key. Unfortunately it's not possible to remove it as the Config component does not work like this.

If you want, you can open a new issue about supporting "prefix" during the import. Something like that:
```yml
# config.yml
imports:
    - { resource: parameters.yml }
    - { resource: security.yml }
    - { resource: services.yml }
    - { resource: workflows.yml, prefix: 'framework.workflow' }
--------------
# workflows.yml
user_process:
    type: 'workflow'
    marking_store:
    #......
```

Next, it looks like it's an issue to declare the configuration in yaml. Like any others configurations in  Symfony, it's possible to declare the configuration in [xml](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/workflows.xml) or plain [PHP](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/php/workflows.php). 

That's why I'm closing this issue.

Thanks.

**Comment by Guikingone** [comment]
Hi @lyrixx, 

Yes, that what i saw trying to build the workflowbundle inside the SE, probably not easy but possible, the injection of the key is probably the hardest way and could cost time into the development. 

For the framework, i'm still convinced that we can get rid of the framework key but that's my personal vision and i work on this way on the fork that i did, for the declaration, i'm wasn't thinking about .yml only but about "global" declaration of configuration about the workflow, i've talk about yaml because that's what i'm using at this time.

**Comment by lyrixx** [comment]
> Yes, that what i saw trying to build the workflowbundle inside the SE, probably not easy but possible, the injection of the key is probably the hardest way and could cost time into the development.

I'm not sure to understand, but even if you have a workflow bundle, the `workflow` key will be here.
So here we are arguing about one key, and an extra indentation of 4 spaces...

> For the framework, i'm still convinced that we can get rid of the framework key but that's my personal vision and i work on this way on the fork that i did, for the declaration, i'm wasn't thinking about .yml only but about "global" declaration of configuration about the workflow, i've talk about yaml because that's what i'm using at this time.

If you have something to review, I will be happy to do it ;)

**Comment by Guikingone** [comment]
The main problem wasn't about the workflow key, problem was about the framework key who's not on the right place with workflow (or the inverse, workflow should be separated from the framework key in my opinion) but yeah, for sure, once i've something who can stand, definitively gonna give you a deeper access :)

**Comment by stof** [comment]
having a prototyped array node as top-level configuration of a bundle is a bad idea, as it makes it very hard to evolve the config in a BC way in the future (thus I'm not even sure it works fine).
so even if bundles were configured by a WorkflowBundle rather than FrameworkBundle, doing this would be a bad idea:

```yaml
workflow:
  my_first_workflow:
    # ...
  my_second_workflow:
    # ...
```

A future-proof way for the bundle would have to be:

```yaml
workflow:
  workflows:
    my_first_workflow:
      # ...
    my_second_workflow:
      # ...
```

**Comment by Guikingone** [comment]
Yes, definitively looking in this direction @stof, the second option is the idea that i have when trying to build the process into the WorkflowBundle.

**Comment by stof** [comment]
Well, how is the second option better than this from a config PoV ?

```yaml
framework:
  workflows:
    my_first_workflow:
      # ...
    my_second_workflow:
      # ...
```

The only difference is the first line.

Then, saying that FrameworkBundle should not be responsible for the configuration of workflows in favor of a different bundle is a separate topic than the one you brought here (which may be a good idea for symfony/flex)

**Comment by Guikingone** [comment]
The first problem was about the workflow key who's located on the config.yml file, in my opinion, that's a bad idea for future evolutions and dev process, it put too much config key into one file, like @lyrixx says, you can put the key into a separate file and inject this last one into the config.yml.

Even if it's a good idea for "dispatching" the logic injected into config.yml, that's a bad idea in the fact that workflow keys depend on framework primary key, for what i've test and do at this time, it's possible to make the workflow key primary has you show into the second example, a little bit harder that the actual configuration but possible. 

It's not the principal problem for what this issue was created but i think that could lead to a reflection or a new vision about the config keys and the coupling of logic into a single file, for what it cost, i think that we could look into this direction.

**Comment by stof** [comment]
well, each bundle has its own top-level config key. So using a different top-level key requires putting things in a separate bundle (which also has drawbacks).

**Comment by SulivanDotEu** [comment]
You can achieve this well easily thanks to Symfony's kernel.

```php
// Kernel.php
<?php
//...
class Kernel extends BaseKernel
{
 
    protected function configureContainer(ContainerBuilder $container, LoaderInterface $loader)
    {
        // ...
        // read multiple files for workflows
        $loader->load($confDir.'/{workflow}/*'.self::CONFIG_EXTS, 'glob');
        // it works well with easy admin too
        $loader->load($confDir.'/{admin}/*'.self::CONFIG_EXTS, 'glob');
        // ...
    }
}
```
